File: packages/talon/bench/bench_talon.ml
(*---------------------------------------------------------------------------
  Copyright (c) 2026 The Raven authors. All rights reserved.
  SPDX-License-Identifier: ISC
  ---------------------------------------------------------------------------*)

(* Talon dataframe benchmarks using realistic CSV fixtures. *)

module Row = Talon.Row

module Fixtures = struct
  let data_dir = Filename.concat (Sys.getcwd ()) "talon/bench/data"

  let load_csv name dtype_spec =
    Talon_csv.read ~dtype_spec (Filename.concat data_dir name)

  let transactions =
    lazy
      (load_csv "transactions.csv"
         [
           ("transaction_id", `Int32);
           ("customer_id", `Int32);
           ("region", `String);
           ("category", `String);
           ("channel", `String);
           ("amount", `Float64);
           ("quantity", `Int32);
           ("discount", `Float64);
           ("promo", `String);
           ("event_date", `String);
         ])

  let customers =
    lazy
      (load_csv "customers.csv"
         [
           ("customer_id", `Int32);
           ("segment", `String);
           ("region", `String);
           ("status", `String);
           ("loyalty_score", `Float64);
           ("tenure_years", `Int32);
         ])

  let transactions () = Lazy.force transactions
  let customers () = Lazy.force customers
end

let force_float_sum df column =
  let total = Talon.Agg.Float.sum df column in
  ignore total

let bench_filter df =
  let filtered =
    Talon.filter_by df
      Row.(
        map3 (float64 "amount") (int32 "quantity") (string "region")
          ~f:(fun amount quantity region ->
            amount > 120.
            && Int32.compare quantity 3l >= 0
            && String.equal region "EMEA"))
  in
  force_float_sum filtered "amount"

let bench_group df =
  let groups =
    Talon.group_by df
      Row.(
        map2 (string "category") (string "region") ~f:(fun category region ->
            category ^ "|" ^ region))
  in
  let total =
    List.fold_left
      (fun acc (_key, group_df) -> acc +. Talon.Agg.Float.sum group_df "amount")
      0. groups
  in
  ignore total

let bench_join df customers =
  let joined = Talon.join df customers ~on:"customer_id" ~how:`Left () in
  force_float_sum joined "amount"

let bench_sort df =
  let sorted = Talon.sort_values ~ascending:false df "amount" in
  force_float_sum sorted "amount"

let all_benchmarks =
  let transactions = Fixtures.transactions () in
  let customers = Fixtures.customers () in
  [
    Ubench.bench "Filter/high_value" (fun () -> bench_filter transactions);
    Ubench.bench "Group/category_region" (fun () -> bench_group transactions);
    Ubench.bench "Join/customer_lookup" (fun () ->
        bench_join transactions customers);
    Ubench.bench "Sort/amount_desc" (fun () -> bench_sort transactions);
  ]
  |> fun benches -> [ Ubench.group "Talon" benches ]

let () = Ubench.run_cli all_benchmarks

File: packages/talon/bench/bench_talon.py
from __future__ import annotations

import sys
from pathlib import Path
from typing import Any, List

import pandas as pd

_ROOT = Path(__file__).resolve().parents[2]
_UBENCH = _ROOT / "vendor" / "ubench"
if str(_UBENCH) not in sys.path:
    sys.path.insert(0, str(_UBENCH))

import ubench  # type: ignore


DATA_DIR = Path(__file__).resolve().parent / "data"


def _load_data() -> tuple[pd.DataFrame, pd.DataFrame]:
    transactions = pd.read_csv(
        DATA_DIR / "transactions.csv",
        dtype={
            "transaction_id": "int32",
            "customer_id": "int32",
            "region": "category",
            "category": "category",
            "channel": "category",
            "amount": "float64",
            "quantity": "int32",
            "discount": "float64",
            "promo": "category",
            "event_date": "string",
        },
    )
    customers = pd.read_csv(
        DATA_DIR / "customers.csv",
        dtype={
            "customer_id": "int32",
            "segment": "category",
            "region": "category",
            "status": "category",
            "loyalty_score": "float64",
            "tenure_years": "int32",
        },
    )
    return transactions, customers


TRANSACTIONS, CUSTOMERS = _load_data()


def build_benchmarks() -> List[Any]:
    benches: List[Any] = []

    def bench_filter() -> None:
        filtered = TRANSACTIONS[
            (TRANSACTIONS["amount"] > 120.0)
            & (TRANSACTIONS["quantity"] >= 3)
            & (TRANSACTIONS["region"] == "EMEA")
        ]
        float(filtered["amount"].sum())

    benches.append(ubench.bench("Filter/high_value (pandas)", bench_filter))

    def bench_group() -> None:
        grouped = (
            TRANSACTIONS.groupby(["category", "region"])["amount"].sum()
        )
        float(grouped.sum())

    benches.append(ubench.bench("Group/category_region (pandas)", bench_group))

    def bench_join() -> None:
        joined = TRANSACTIONS.merge(CUSTOMERS, on="customer_id", how="left")
        float(joined["amount"].sum())

    benches.append(ubench.bench("Join/customer_lookup (pandas)", bench_join))

    def bench_sort() -> None:
        sorted_df = TRANSACTIONS.sort_values("amount", ascending=False)
        float(sorted_df["amount"].iloc[0])

    benches.append(ubench.bench("Sort/amount_desc (pandas)", bench_sort))

    return benches


def default_config() -> ubench.Config:
    return ubench.Config.default().build()


def main() -> None:
    benchmarks = build_benchmarks()
    config = default_config()
    ubench.run(benchmarks, config=config, output_format="pretty", verbose=False)


if __name__ == "__main__":
    main()
